File: backend/src/main/java/com/example/backend/service/FriendshipService.java
package com.example.backend.service;

import com.example.backend.model.respon.FriendResponse;
import com.example.backend.model.respon.FriendshipResponse;

public interface FriendshipService {
    // gửi lời mời kết bạn
    FriendshipResponse sendFriendRequest(Long senderId, Long receiverId);

    // chấp nhận lời mời kết bạn
    FriendResponse acceptFriendRequest(Long receiverId, Long senderId);

    // từ chối lời mời kết bạn
    Boolean rejectFriendRequest(Long receiverId, Long senderId);

    // hủy lời mời kết bạn đã gửi
    Boolean cancelFriendRequest(Long senderId, Long receiverId);

    // hủy kết bạn
    Boolean unfriend(Long userId, Long friendId);

    // chặn người dùng
    Boolean blockUser(Long blockerId, Long blockedId);

    // bỏ chặn người dùng
    Boolean unblockUser(Long blockerId, Long blockedId);

    // lấy danh sách bạn bè
    java.util.List<FriendResponse> getFriends(Long userId);

    // lấy danh sách lời mời kết bạn (pending requests)
    org.springframework.data.domain.Page<FriendshipResponse> getPendingRequests(Long userId, int page, int size);

    // lấy danh sách người dùng đã bị chặn
    java.util.List<FriendResponse> getBlockedUsers(Long userId);
}

File: frontend/src/service/impl/FriendShipService.ts
import { http } from "../http/http";
import type { ApiDataResponse, Page, FriendResponse, FriendshipResponse } from "../../utils/ResponType";

export const getFriends = (currentUserId: number) => {
    return http.get<ApiDataResponse<FriendResponse[]>>(`/friends?currentUserId=${currentUserId}`);
};

export const sendFriendRequest = (currentUserId: number, receiverId: number) => {
    return http.post<ApiDataResponse<FriendshipResponse>>(`/friends/requests/${receiverId}?currentUserId=${currentUserId}`);
};

export const acceptFriendRequest = (currentUserId: number, senderId: number) => {
    return http.post<ApiDataResponse<FriendResponse>>(`/friends/requests/${senderId}/accept?currentUserId=${currentUserId}`);
};

export const rejectFriendRequest = (currentUserId: number, senderId: number) => {
    return http.post<ApiDataResponse<boolean>>(`/friends/requests/${senderId}/reject?currentUserId=${currentUserId}`);
};

export const cancelFriendRequest = (currentUserId: number, receiverId: number) => {
    return http.delete<ApiDataResponse<boolean>>(`/friends/requests/${receiverId}?currentUserId=${currentUserId}`);
};

export const unfriend = (currentUserId: number, friendId: number) => {
    return http.delete<ApiDataResponse<boolean>>(`/friends/${friendId}?currentUserId=${currentUserId}`);
};

export const blockUser = (currentUserId: number, blockedId: number) => {
    return http.post<ApiDataResponse<boolean>>(`/friends/${blockedId}/block?currentUserId=${currentUserId}`);
};

export const unblockUser = (currentUserId: number, blockedId: number) => {
    return http.post<ApiDataResponse<boolean>>(`/friends/${blockedId}/unblock?currentUserId=${currentUserId}`);
};

export const getPendingRequests = (currentUserId: number, page: number = 0, size: number = 20) => {
    return http.get<ApiDataResponse<Page<FriendshipResponse>>>(
        `/friends/requests/pending?currentUserId=${currentUserId}&page=${page}&size=${size}`
    );
};

export const getBlockedUsers = (currentUserId: number) => {
    return http.get<ApiDataResponse<FriendResponse[]>>(`/friends/blocked?currentUserId=${currentUserId}`);
};
